src/node_report.cc

```diff
@@ -0,0 +1,1115 @@
+using v8::RegisterState;
+using v8::SampleInfo;
+using v8::String;
+using v8::V8;
+
+// Internal/static function declarations
+static void WriteNodeReport(Isolate* isolate, DumpEvent event,
+                            const char* message, const char* location,
+                            char* filename, std::ostream &out,
+                            MaybeLocal<Value> error, TIME_TYPE* time);
+static void PrintCommandLine(std::ostream& out);
+static void PrintVersionInformation(std::ostream& out);
+static void PrintJavaScriptStack(std::ostream& out, Isolate* isolate,
+                                 DumpEvent event, const char* location);
+static void PrintJavaScriptErrorStack(std::ostream& out, Isolate* isolate,
+                                      MaybeLocal<Value> error);
+static void PrintStackFromStackTrace(std::ostream& out, Isolate* isolate,
+                                     DumpEvent event);
+static void PrintStackFrame(std::ostream& out, Isolate* isolate,
+                            Local<StackFrame> frame, int index, void* pc);
+static void PrintNativeStack(std::ostream& out);
+#ifndef _WIN32
+static void PrintResourceUsage(std::ostream& out);
+#endif
+static void PrintGCStatistics(std::ostream& out, Isolate* isolate);
+static void PrintSystemInformation(std::ostream& out, Isolate* isolate);
+static void PrintLoadedLibraries(std::ostream& out, Isolate* isolate);
+
+// Global variables
+static int seq = 0;  // sequence number for report filenames
+const char* v8_states[] = {"JS", "GC", "COMPILER", "OTHER", "EXTERNAL", "IDLE"};
+static bool report_active = false;  // recursion protection
+char report_filename[NR_MAXNAME + 1] = "";
+char report_directory[NR_MAXPATH + 1] = "";  // default: current directory
+TIME_TYPE loadtime_tm_struct;  // module load time
+time_t load_time;  // module load time absolute
+
+// std::string version_string = NODE_VERSION_STRING;
+// std::string commandline_string = "";
+version_and_command_struct version_and_command;
+
+
+
+/*******************************************************************************
+ * External function to trigger a node report, writing to file.
+ *
+ * The 'name' parameter is in/out: an input filename is used if supplied, and
+ * the actual filename is returned.
+ ******************************************************************************/
+void TriggerNodeReport(Isolate* isolate, DumpEvent event, const char* message,
+                       const char* location, char* name,
+                       MaybeLocal<Value> error) {
+  // Recursion check for report in progress, bail out
+  if (report_active) return;
+  report_active = true;
+
+  // Obtain the current time and the pid (platform dependent)
+  TIME_TYPE tm_struct;
+#ifdef _WIN32
+  GetLocalTime(&tm_struct);
+  DWORD pid = GetCurrentProcessId();
+#else  // UNIX, OSX
+  struct timeval time_val;
+  gettimeofday(&time_val, nullptr);
+  localtime_r(&time_val.tv_sec, &tm_struct);
+  pid_t pid = getpid();
```

Comment:
`uv_os_getpid` works cross-platform (and it might be nice to eventually do something consistent for the time as well)

Reply:
done

Reply:
for `time`, will do some research.